Adding NCL_lcmask_1 example file

hCraker opened the pull request:
Closes issue #59

hCraker commented:
I have the map showing and the colorbar created, but I'm unable to plot the data on the map. I'm not sure what the issue is. Can anyone offer some advice?
@jukent @erogluorhan @kmpaul 
![image](https://user-images.githubusercontent.com/42781301/82708985-98e85800-9c4d-11ea-8758-a9455d9a0b31.png)

jukent commented:
This is interesting. The code doesn't seem to be failing at the plotting step since it produces the colorbar (which you specify afterwards). I will try to reproduce the error and find a fix (but it might not be until I am back "in the office" on Tuesday.

hCraker commented:
@jukent I realized that I had forgotten to transform the data into the correct map projection. I've solved my original issue, and I've moved on to implementing the other features of the map.
![image](https://user-images.githubusercontent.com/42781301/82950808-fcd38f00-9f73-11ea-9478-0a07a7eec24c.png)

hCraker commented:
![image](https://user-images.githubusercontent.com/42781301/83063821-ba22bd00-a02e-11ea-91b6-71b8a3e3fbf3.png)

jukent commented:
@hCraker This looks great!

I am a little confused about the issue you described making the lcmask_1_2_lg.png plot conical, since the first plot you posted is conical. Maybe ping me once you start working on that more.

hCraker commented:
@jukent When I specify the bounds for the map (i.e. between 40W and 30E) the plot becomes a rectangle. I have not yet figured out how to make it appear conic, but I have only just started working on that issue. I'll ping you when I push a commit with the solution.

jukent commented:
Yes it will be easier for me to help when the code for that map is up too.

hCraker commented:
Figure lcmask_1_1_lg
![image](https://user-images.githubusercontent.com/42781301/83195708-a8ade380-a108-11ea-896b-443ec55b3e30.png)

Figure lcmask_1_2_lg
![image](https://user-images.githubusercontent.com/42781301/83195784-c7ac7580-a108-11ea-9780-b446fa0c1342.png)

This is what I have so far. I've noticed that in the second example plot, the NCL version has the longitude shifted by 180 degrees. I find this odd because there isn't a practical purpose for the translation, but it explains why my data looks different for lcmask_1_2_lg.

I'm still working to figure out how to refine the shape of the lcmask_1_2_lg plot.

hCraker commented:
The commit that I just pushed has the following line added 
`ax.set_extent([-90, 40, 20, 80], ccrs.PlateCarree())`
I was trying to zoom into the region 270W-40E 20N-80N, but the bottom of the plot is cut off. This is odd because the curve on the plot from the commit before is 20N. Does anyone know a better way to zoom into an area that doesn't result in a rectangular plot? @jukent @erogluorhan @clyne @kmpaul 

This is what the code currently produces.
![image](https://user-images.githubusercontent.com/42781301/83303040-651bae00-a1ca-11ea-85c4-5544e1a527bd.png)

jukent commented:
Have you tried changing the `ccrs.PlateCarree()`, since that is a rectangular projection, to the projection you are plotting?

hCraker commented:
> Have you tried changing the `ccrs.PlateCarree()`, since that is a rectangular projection, to the projection you are plotting?

@jukent I have tried that and this is the result. From the documentation I've read, it suggests that the projection argument for `ax.set_extent` is to specify the projection of the coordinates argument. I wonder if I need to transform the coordinates into the Lambert Conformal projection instead, but I'm not sure how to do that. I'm looking into it.

![image](https://user-images.githubusercontent.com/42781301/83423263-b9a47080-a3f8-11ea-84ef-030000630cd2.png)

jukent commented:
@hCraker I wonder if you have to do something similar to this example: https://scitools.org.uk/cartopy/docs/latest/gallery/always_circular_stereo.html#sphx-glr-gallery-always-circular-stereo-py

```
    # Compute a circle in axes coordinates, which we can use as a boundary
    # for the map. We can pan/zoom as much as we like - the boundary will be
    # permanently circular.
    theta = np.linspace(0, 2*np.pi, 100)
    center, radius = [0.5, 0.5], 0.5
    verts = np.vstack([np.sin(theta), np.cos(theta)]).T
    circle = mpath.Path(verts * radius + center)

    ax2.set_boundary(circle, transform=ax2.transAxes)
```
except instead of a circle you would define the shape of the cone. 

@clyne Custom boundary shape in Cartopy seems like something GeoCAT could make into a cleaner and easier function.

hCraker commented:
Thank you @jukent I will try something similar.

jukent commented:
Here is another example of making a star-shaped map:
https://scitools.org.uk/cartopy/docs/latest/gallery/star_shaped_boundary.html#sphx-glr-gallery-star-shaped-boundary-py

Maybe matplotlib.mpath has a conical shape?

hCraker commented:
I have managed to make a wedge shaped boundary; however, there is a stray line coming from the boundary that I'm having trouble removing. Also, the bounding box for the plot axes is too large.

![image](https://user-images.githubusercontent.com/42781301/83452721-3cdcbb00-a427-11ea-9c9e-a32e62e181ed.png)

jukent commented:
Which methods did you try?

hCraker commented:
Below is what I added. I also replaced the line with `ax.set_extent()` to `ax.set_global()`
```
center, radius = [0.5, 0.5], 0.5
theta1 = 210
theta2 = 332.5
wedge = mpatches.Wedge(center, radius, theta1, theta2, width=0.43, transform=ax.transAxes).get_path()
ax.set_boundary(wedge, transform=ax.transAxes)
```

jukent commented:
Perhaps try creating your own wedge, more similar to the circle example - to see if that is part of the problem.

hCraker commented:
I created a custom boundary that looks much better, but my additions could likely be made into a cleaner function in the future. Should I make an issue for that in this repo or would it be better suited for geocat-viz @clyne @jukent ?

I'm going to attempt to fix the bounding box next. I think the issue stems from reducing the visible area of the axes, but not the actual space they take up.

![image](https://user-images.githubusercontent.com/42781301/83459482-77982080-a432-11ea-9931-7f8acb56d270.png)

clyne commented:
> I created a custom boundary that looks much better, but my additions could likely be made into a cleaner function in the future. Should I make an issue for that in this repo or would it be better suited for geocat-viz @clyne @jukent ?
> 
> I'm going to attempt to fix the bounding box next. I think the issue stems from reducing the visible area of the axes, but not the actual space they take up.
> 
> ![image](https://user-images.githubusercontent.com/42781301/83459482-77982080-a432-11ea-9931-7f8acb56d270.png)

How about seeing if you can turn your specialized code for the "wedge" into a generalized helper function, that possibly takes an inner and outer radius as arguments, along with a starting and stopping arc angle? That would be a good step to fleshing out any issues.

hCraker commented:
@clyne I have added a general function for drawing the wedge shape. The plots looks the same, and I have been unable to figure out how to reduce the amount of white space in the plot.

hCraker commented:
@jukent @clyne I have tried a slightly different method to create the wedge shaped boundary, and this is the result. I have a feeling I'm getting closer, but I still don't know how to get this exactly right. Any advice would be greatly appreciated.

![image](https://user-images.githubusercontent.com/42781301/83572541-c8258180-a4f7-11ea-908e-fbbdfd5bb0fe.png)

clyne commented:
> @clyne I have added a general function for drawing the wedge shape. The plots looks the same, and I have been unable to figure out how to reduce the amount of white space in the plot.

The utility function looks great. Nice job! Re the white space: I would guess, and it would only be a guess, that the set_boundary method doesn't change the extents of the bounding box. You may have to do that manually.

jukent commented:
@hCraker What have you done to try to remove the whitespace? Does `plt.ylim` do the trick?

hCraker commented:
I think I'm nearly done. Here are what the plots look like. Please let me know if you have any suggestions.

![image](https://user-images.githubusercontent.com/42781301/83693475-24ee6e00-a5c4-11ea-91a2-f223354b80e9.png)
![image](https://user-images.githubusercontent.com/42781301/83693488-2b7ce580-a5c4-11ea-80d2-9d6c44367a67.png)

clairefio commented on Plots/Masking/NCL_lcmask_1.py:
```diff
@@ -0,0 +1,144 @@
+wind = masked.plot.contourf(ax=ax, cmap=newcmp, transform=ccrs.PlateCarree(),
+                            add_colorbar=False, levels=24)
+plt.colorbar(wind, ax=ax, orientation='horizontal', drawedges=True,
+             ticks=np.arange(-40, 44, 4), pad=0.1, aspect=18)
```
The original NCL plot doesn't have colorbar ticks

clairefio commented on Plots/Masking/NCL_lcmask_1.py:
```diff
@@ -0,0 +1,144 @@
+wind = V.plot.contourf(ax=ax, cmap=newcmp, transform=ccrs.PlateCarree(),
+                       add_colorbar=False, levels=24)
+plt.colorbar(wind, ax=ax, orientation='horizontal', drawedges=True,
+             ticks=np.arange(-48, 48, 8), pad=0.1, aspect=12)
```
The original NCL plot doesn't have colorbar ticks

erogluorhan commented on Plots/Masking/NCL_lcmask_1.py:
```diff
@@ -0,0 +1,147 @@
+# xarrays and disable time decoding due to missing necessary metadata
+ds = xr.open_dataset(gdf.get("netcdf_files/atmos.nc"), decode_times=False)
+# Extract a slice of the data
+ds = ds.isel(time=0).drop("time")
```
There is a future deprecation warning for using `drop` function for dropping labels and it suggests to use `drop_sel` instead if possible. Please try to see the deprecation warning on your own and it would be good to avoid such a deprecation before merging the code.

hCraker replied:
`drop_vars` is a function recommended on the xarray page for `drop`. I have used that instead and it works.

erogluorhan commented on Plots/Masking/NCL_lcmask_1.py:
```diff
@@ -0,0 +1,147 @@
+cbar.ax.tick_params(length=0) # remove tick marks but leave in labels
+
+# Add titles
+plt.title(V.long_name, loc='left', size=16)
```
How about using `gvutil.set_titles_and_labels()` instead of these two lines?

erogluorhan commented on Plots/Masking/NCL_lcmask_1.py:
```diff
@@ -0,0 +1,147 @@
+cbar.ax.tick_params(length=0) # remove tick marks but leave in labels
+
+# Add titles
+plt.title(masked.long_name, loc='left', size=12)
```
Same as above.

hCraker commented on Plots/Masking/NCL_lcmask_1.py:
```diff
@@ -0,0 +1,147 @@
+===============
+This script illustrates the following concepts:
+   - Drawing filled contours over a Lambert Conformal map
+   - Drawing a filled contours over a masked Lambert Conformal plot
```
Good catch! It is redundant, and I've fixed it

hCraker commented:
> Great work and it is ready for being merged. Though, let us keep it open until `set_wedge_boundary` function is merged to **geocat-viz** with the help of PR [9](https://github.com/NCAR/geocat-viz/pull/9). That function handles the same task as the function `wedge_boundary` does here, right? IF so, whenever we merge geocat-viz, we can remove function `wedge_boundary` from here and just call it from **geocat-viz** instead.

That sounds like a good idea to me. I mentioned this PR in PR [9](https://github.com/NCAR/geocat-viz/pull/9) to make sure we remember to modify this example.

erogluorhan commented:
> That sounds like a good idea to me. I mentioned this PR in PR [9](https://github.com/NCAR/geocat-viz/pull/9) to make sure we remember to modify this example.

This would work well, too (i.e. mentioning this PR in PR [9](https://github.com/NCAR/geocat-viz/pull/9)). May be we can do this: If there was any other GeoCAT-examples PRs that also implemented `wedge_boundary` function on its own previously, can you put all of such functions' links including this PR to the initial description comment of **geocat-viz** PR [9](https://github.com/NCAR/geocat-viz/pull/9))? This is because mentioning appears well below the conversations list and can likely get overlooked, but we can always see related PR links in the initial comment.

Does that make sense? 

Fyi: I will have a look at PR [9](https://github.com/NCAR/geocat-viz/pull/9) today and if there is much review process needed there, we can merge this PR here and then revisit whenever PR [9](https://github.com/NCAR/geocat-viz/pull/9) is ok.

hCraker commented:
> This would work well, too (i.e. mentioning this PR in PR [9](https://github.com/NCAR/geocat-viz/pull/9)). May be we can do this: If there was any other GeoCAT-examples PRs that also implemented `wedge_boundary` function on its own previously, can you put all of such functions' links including this PR to the initial description comment of **geocat-viz** PR [9](https://github.com/NCAR/geocat-viz/pull/9))? This is because mentioning appears well below the conversations list and can likely get overlooked, but we can always see related PR links in the initial comment.
> 
> Does that make sense?
> 
> Fyi: I will have a look at PR [9](https://github.com/NCAR/geocat-viz/pull/9) today and if there is much review process needed there, we can merge this PR here and then revisit whenever PR [9](https://github.com/NCAR/geocat-viz/pull/9) is ok.

This [issue](https://github.com/NCAR/GeoCAT-examples/issues/91) in **GeoCAT-examples** is mentioned at the beginning of PR [9](https://github.com/NCAR/geocat-viz/pull/9) in **GeoCAT-viz**. It was made to make sure we fix the examples that implement the wedge boundary on their own. Will that suffice?

erogluorhan commented:
> > This would work well, too (i.e. mentioning this PR in PR [9](https://github.com/NCAR/geocat-viz/pull/9)). May be we can do this: If there was any other GeoCAT-examples PRs that also implemented `wedge_boundary` function on its own previously, can you put all of such functions' links including this PR to the initial description comment of **geocat-viz** PR [9](https://github.com/NCAR/geocat-viz/pull/9))? This is because mentioning appears well below the conversations list and can likely get overlooked, but we can always see related PR links in the initial comment.
> > Does that make sense?
> > Fyi: I will have a look at PR [9](https://github.com/NCAR/geocat-viz/pull/9) today and if there is much review process needed there, we can merge this PR here and then revisit whenever PR [9](https://github.com/NCAR/geocat-viz/pull/9) is ok.
> 
> This [issue](https://github.com/NCAR/GeoCAT-examples/issues/91) in **GeoCAT-examples** is mentioned at the beginning of PR [9](https://github.com/NCAR/geocat-viz/pull/9) in **GeoCAT-viz**. It was made to make sure we fix the examples that implement the wedge boundary on their own. Will that suffice?

Yes, thanks!
